lib/HeroBonus.cpp

```diff
@@ -427,6 +427,9 @@ int BonusList::totalValue() const
 
 	for(std::shared_ptr<Bonus> b : bonuses)
 	{
+		if(b->effectRange == Bonus::ONLY_ENEMY_ARMY)

```

Comment:
This only fixes bonuses with value, but what about things like NO_MORALE

Reply:
Also, `anaffectedByMoraleSelector` has been fixed.

Reply:
It was just an example to show that you need to ignore not only bonus value calculation but also presence of bonus.

Reply:
You should check `IBonusBearer::hasBonus`, also in `IBonusBearer::hasBonus(const CSelector &selector, const CSelector &limit, const std::string &cachingStr)`  overload second parameter `limit` is specifically for effectRange ,

Reply:
As far as I understand, you propose add 'effectRange' processing to the 'hasBonus' code to hide presence of bonuses intended for enemy army.  It would be some kind of hack and it's out of this PR scope. 

'effectRange' processing is related to the game business logic of a higher level, than the bonus propagation and bonus limiters logic. The bonus limiters work at the level of node tree access, and the 'effectRange' works at the level of the game business logic. So, no need to mix high-level game logic and low-level node tree data access logic. 'hasBonus' methods are a kind of data access API. Business logic should have no place in the data access code. 

Additionally, "effectRange = ONLY_ENEMY_ARMY" doesn't make sense for most of the bonuses without value, so problem you propose to solve is not actual at all.

Reply:
"effectRange = ONLY_ENEMY_ARMY" is a hack on its own. It should be implemented with propagation (BATTLE_WIDE propagator and a limiter for opposite side of bonus source - the tricky part nobody managed to implement yet)

Reply:
I am inclined to conclude, that this does not have a natural solution within the framework of the bonus system. 

The creature bonus affects all players. Therefore, when creating a stack with a given creature, the bonus from the creature must be customized with information about the current owner. At this stage, we should create a copy of the creature bonus and add information about the current owner to this copy to avoid affecting other players. In this implementation, the bonus on the creature is idle and serves only to create a real bonus. It is better if that idle bonus from creature would not have a propagator, otherwise the BATTLE node will receive an extra bonus from creature node with owner=neutral.

Thus, we get the same two-stage propagation scheme. In this case, in JSON, it would still be better to use effectRange with an empty propagator and then create the desired propagator with the desired limiter. Btw, effectRange will have to be supported anyway for backward compatibility reasons.

On the other hand, there is a separate issue related to the  bonus propagation from the city to the visitor or to the defending hero, which is solved in a different way.

Thus, the logic of the solution becomes blurred, the benefits are not visible, no improvements are foreseen for modders (the corresponding record will be the same or even longer).

And the architectural reasons for this are obvious - an attempt to bring the high-level concept of "current owner - player" to the low level of the node tree and the distribution of bonuses.

Reply:
So, @alexvins , what do you think about it?

Reply:
I do not like the two-stage scheme, but see no more issues preventing it to be merged.